### PR TITLE
ucp 7058: raftstore: pause split check when bulk insert

### PR DESCRIPTION
**Pull request by SSebo**
UCP #7058 

Signed-off-by: ssebo <ssebo.zzz@gmail.com>

### What problem does this PR solve?

When bulk insert is too fast, a region can be split frequently, which will make snapshot stale very soon. It's possible that a follower can hardly catch up logs with leader when every snapshot becomes stale immediately

### What is changed and how it works?

What's Changed:

- add `skip_split_count` in `fsm/peer.rs` 
- add `fn is_generating_snapshot()` in `peer_storage`

How it Works:

There is a trade off between region size and snapshot success rate. Split check is triggered every 10 seconds, If a snapshot can't be generated in 30 seconds, it might be just too large to be generated. Split it into smaller size can help generation.
### Related changes

None

### Check List 

Tests 
- Integration test

###  Refer to a related PR or issue link (optional)
https://github.com/pingcap/ticdc/issues/330

### Side effects

- Performance regression

- Breaking backward compatibility

### Release note

### CODE REVIEW

**Comment by SSebo** [comment]
@BusyJay mentor, skip is added, but no idea how to construct a test case, any suggestion ?

**Comment by BusyJay** [comment]
You can utilize `region_gen_snap` failpoint to simulate long snapshot generation, and see if a region is not split in a check interval.

**Comment by SSebo** [comment]
@BusyJay mentor
- I tried to use `region_gen_snap` to hold the snap generating state, but the state is still changing over time when I print it.
- I add one fail point `is_generating_snapshot`  holding the state to simulate huge snap generation
- I add another fail point `region_split_skip_max_count` to make skip times very large to never split. I found small value not stable in the test.

File: components/raftstore/src/store/fsm/peer.rs [hunk 1]
```diff
@@ -2528,6 +2532,12 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
         )
     }
 
+    #[inline]
+    fn region_split_skip_max_count(&self) -> u8 {
+        fail_point!("region_split_skip_max_count", |_| { u8::max_value() });
+        2
```
Comment (BusyJay) [inline components/raftstore/src/store/fsm/peer.rs]: I think it should be 3.

File: tests/failpoints/cases/test_split_region.rs [hunk 2]
```diff
@@ -63,6 +67,68 @@ fn test_follower_slow_split() {
...
+    assert_ne!(left, right);
+    assert_eq!(region.get_start_key(), left.get_start_key());
+    assert_eq!(right.get_start_key(), left.get_end_key());
+    assert_eq!(region.get_end_key(), right.get_end_key());
+
+    fail::remove(is_generating_snapshot);
+}
+
+#[test]
+fn test_pause_split_when_snap_gen_never_split() {
```
Comment (BusyJay) [inline tests/failpoints/cases/test_split_region.rs]: The case can run too long, better use smaller timeout.

File: components/raftstore/src/store/fsm/peer.rs (hunk 1 above)
Comment (SSebo) [reply components/raftstore/src/store/fsm/peer.rs]: My bad, It should be 3.

File: tests/failpoints/cases/test_split_region.rs (hunk 2 above)
Comment (SSebo) [reply tests/failpoints/cases/test_split_region.rs]: I'll change to try 50 times, 20ms timeout, which is 1 second.

**Comment by SSebo** [comment]
@BusyJay mentor, split time change to 3, test case longest waiting change to 1s. PTAL

File: tests/failpoints/cases/test_split_region.rs (hunk 2 above)
Comment (BusyJay) [reply tests/failpoints/cases/test_split_region.rs]: 1 second is still too long. You can adapt `gen_split_region` to accept custom timeout. 200ms should be enough for one split.

Comment (SSebo) [reply tests/failpoints/cases/test_split_region.rs]: got it.

**Comment by SSebo** [comment]
@BusyJay mentor, I change base tick to 20ms, and wait will try 20 times, each 10 ms. PTAL.

**Comment by Fullstop000** [comment]
@SSebo `test_search` will be fixed by https://github.com/tikv/tikv/pull/7273. You can just ignore it.  Sorry for the bothering

**Comment by SSebo** [comment]
@Fullstop000 Thanks for your notify.

**Comment by SSebo** [comment]
@BusyJay mentor, all tests passed, PTAL.

**Comment by SSebo** [comment]
@BusyJay mentor, need two approval to merge, @youjiali1995 Could you help to review this PR?

File: components/raftstore/src/store/fsm/peer.rs [hunk 3]
```diff
@@ -2525,6 +2529,12 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
         )
     }
 
+    #[inline]
+    fn region_split_skip_max_count(&self) -> u8 {
+        fail_point!("region_split_skip_max_count", |_| { u8::max_value() });
+        3
```
Comment (hicqu) [inline components/raftstore/src/store/fsm/peer.rs:2535]: It's better to define a const.

File: components/raftstore/src/store/fsm/peer.rs [hunk 4]
```diff
@@ -98,6 +98,8 @@ pub struct PeerFsm<E: KvEngine> {
     early_apply: bool,
     mailbox: Option<BasicMailbox<PeerFsm<E>>>,
     pub receiver: Receiver<PeerMsg<E>>,
+    /// when snapshot is generating or sending, skip split check at most 3 times.
```
Comment (hicqu) [inline components/raftstore/src/store/fsm/peer.rs:101]: Please don't hard code `3`.

**Comment by hicqu** [comment]
rest LGTM.

**Comment by SSebo** [comment]
@hicqu @BusyJay  mentors, all checks have passed to merge.

**Pull request merged**